### PR TITLE
Path utilities: use pwd for usernames on unix

### LINKED ISSUE #19977: Spack picks up wrong user after ksu
When trying to run an installation as a user that was switched to using `ksu`, Spack still determines the temp directory (set as `$tempdir/$user/spack-stage`) to be the $user from login, not the one switched to with `ksu`.

### Steps to reproduce the issue

```console
$ ksu new-user
$ whoami
new-user
$ spack python -c "import spack.util.path as p; p.canonicalize_path('$tempdir/$user')"
'/tmp'
$ python
>>> import getpass
>>> print(getpass.getuser())
old-user
>>> import pwd
>>> print pwd.getpwuid(os.getuid())
pwd.struct_passwd(pw_name='new-user', pw_passwd='x', pw_uid=#####, pw_gid=#####, pw_gecos='<info>', pw_dir='/home/new-user', pw_shell='/bin/bash')
...
```

### Error Message

No error message, per se.  Just the fact that when building, the build stage ends up under `~/.spack`, which in the case of our HPC machines is an NFS mounted filesystem and causes the builds to be very slow, especially for expansive Spack Stacks installations.

### Information on your system

```console
$ spack debug report
* **Spack:** 0.15.3
* **Python:** 3.6.8
* **Platform:** linux-rhel7-broadwell
```

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [ ] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by becker33**
Fixes #19977 

`getpass.getuser()` does not respect username changes via `ksu`. `pwd.getpwuid(os.getuid())` does.

@nicholas-sly

### CODE REVIEW

**Comment by becker33** [comment]
The untested lines of code will not be exercised until we start testing on windows, all of our test systems support the python `pwd` module.

File: lib/spack/spack/util/path.py [hunk 1]
```diff
@@ -24,10 +24,21 @@
     'substitute_path_variables',
     'canonicalize_path']
 
+
+def get_user():
+    try:
+        # user pwd for unix systems
+        import pwd
+        return pwd.getpwuid(os.getuid()).pw_name
```
Comment (tgamblin) [inline lib/spack/spack/util/path.py]: Two things:

1. This should be outside the try/except, as it's really the import you're checking here.
2. Shouldn't this be `os.geteuid()` for the effective uid? I think this one fails with setuid, doesn't it?

File: lib/spack/spack/util/path.py [hunk 2]
```diff
@@ -24,10 +24,21 @@
     'substitute_path_variables',
     'canonicalize_path']
 
+
+def get_user():
```
Comment (tgamblin) [inline lib/spack/spack/util/path.py:42]: some comment in the function should explain why we do this (i.e., b/c sudo)

**Review by tgamblin — CHANGES_REQUESTED** [review]
Some feedback. I think this could use a test.

**Review by tgamblin — DISMISSED** [review]

**Comment by becker33** [comment]
@spackbot fix style

**Review by tgamblin — APPROVED** [review]

**Pull request merged**